internal/authz/github/github.go

```diff
@@ -228,27 +246,104 @@ func (p *Provider) FetchRepoPerms(ctx context.Context, repo *extsvc.Repository,
 
 	// 100 matches the maximum page size, thus a good default to avoid multiple allocations
 	// when appending the first 100 results to the slice.
-	userIDs := make([]extsvc.AccountID, 0, 100)
+	const userPageSize = 100
+	userIDs := make([]extsvc.AccountID, 0, userPageSize)
+	seenUsers := make(map[extsvc.AccountID]struct{}, userPageSize)
+
+	// addUserToRepoPerms checks if the given users are already tracked before adding it to perms.
+	addUserToRepoPerms := func(users ...extsvc.AccountID) {
+		for _, user := range users {
+			if _, exists := seenUsers[user]; !exists {
+				seenUsers[user] = struct{}{}
+				userIDs = append(userIDs, user)
+			}
+		}
+	}
+
+	// If groups caching is enabled, we sync just direct affiliations, and sync org/team
+	// collaborators separately from cache
+	affiliation := github.AffiliationDirect
+	if p.groupsCache == nil {
+		// Otherwise, sync all affiliations.
+		affiliation = ""
+	}
+
+	// Sync collaborators
 	hasNextPage := true
 	for page := 1; hasNextPage; page++ {
 		var err error
 		var users []*github.Collaborator
-		users, hasNextPage, err = p.client.ListRepositoryCollaborators(ctx, owner, name, page)
+		users, hasNextPage, err = p.client.ListRepositoryCollaborators(ctx, owner, name, page, affiliation)
 		if err != nil {
 			return userIDs, err
 		}
 
 		for _, u := range users {
-			userIDs = append(userIDs, extsvc.AccountID(strconv.FormatInt(u.DatabaseID, 10)))
+			addUserToRepoPerms(extsvc.AccountID(strconv.FormatInt(u.DatabaseID, 10)))
+		}
+	}
+
+	// If groups caching is disabled, we are done.
+	if p.groupsCache == nil {
+		return userIDs, nil
+	}
+
+	// Get groups affiliated with this repo.
+	groups, err := p.getRepoAffiliatedGroups(ctx, owner, name, opts)
+	if err != nil {
+		return userIDs, errors.Wrap(err, "get groups affiliated with repo")
+	}
+
+	// Perform a fresh sync with groups that need a sync.
+	repoID := extsvc.RepoID(repo.ID)
+	for _, group := range groups {
+		if len(group.Users) > 0 {
+			// Just use cache if available and not invalidated
+			addUserToRepoPerms(group.Users...)
+			// If this repo's membership in this group is not noted yet, add it
+			hasRepo := false
+			for _, user := range group.Repositories {
+				if user == repoID {
+					hasRepo = true
+					break
+				}
+			}
+			if !hasRepo {
+				group.Repositories = append(group.Repositories, repoID)
+				p.groupsCache.setGroup(group.cachedGroup)
+			}
+		} else {
```

Comment:
```suggestion
			}
			
			continue
		}
```

Seems we can use a `continue` to eliminate the `else` clause.

Reply:
fa131c2d7499ebcad50cca40278ad42a80397e0d